### PR TITLE
s_server -WWW should open files in binary-mode

### PR DESCRIPTION
**Pull request by lzsiga**
Related issue #8807

### CODE REVIEW

**Comment by paulidale** [comment]
Should the returned content type be changed to something other than `text/plain` (line 3199)?
If so, would this constitute a change in behaviour?

What about adding a command line option that makes the change?

**Comment by richsalz** [comment]
A command-line flag to specify the content-type is as far as I would go down this path.

**Comment by lzsiga** [comment]
Hi, I think `Content-Type` would be another question, much more complicated than this trivial change. To be frank, for _static files_ the decent value for `Content-Type` would be _How am I supposed to know?_

Of course an actual web-server might consult a 'file-extension to mime-type' database, or might use `libmagic` to examine the file-content, but for the `s_server` it would be an overkill. I think its goal is to check TLS-connection: can it transfer files (including large binary files), and how fast. Still  there is `s_server -HTTP` option for special usages.

What I'd suggest (and could provide a patch for) is the following:

    File extension     Content-Type
    (case-insensitive)
    .txt               text/plain
    .htm .html         text/html
    -                  application/octet-stream

PS: Also I don't think you should specify Content-Type from command line as you cannot tell _which file_ will be asked for by the client.

**Comment by richsalz** [comment]
This comment, https://github.com/openssl/openssl/pull/8811#issuecomment-487047768, is *exactly* the thing that should not be added to the simple apps s_server.  Installing any one of free servers is not very difficult.  Don't try to turn s_server into a "real" webserver.

**Comment by lzsiga** [comment]
Hi, richsalz, I do agree with you that s_server is not meant to became an actual web-server,
but function `s_server.c`:`www_body` already has this logic:
```
File-extension      Content-Type
.htm .html .php     text/html
-                   text/plain
```

Here is what I suggest (two options):

1. change nothing (I'd prefer this, as my only intension was to prevent a Windows-specific problem caused by 'text-mode fopen')

2. change the logic to this:

```
File-extension      Content-Type
.txt                text/plain
.htm .html          text/html
-                   application/octet-stream
```

**Comment by levitte** [comment]
Not all operating systems have exclusively stream-oriented files.  Imagine the surprise when suddenly getting this (in od -c form):

```
0000000  \r  000 H   e   l   l   o   ,       w   o   r   l   d   !   000
```

(this is a text record, starts with a 16-bit record length, the text itself, and padding)

**Comment by paulidale** [comment]
@levitte that's the kind of thing I was concerned about.

**Comment by lzsiga** [comment]
Well, I know only Siemens BS2000, it doesn't switch into "raw uninterpreted datablock mode" even in mode="rb", but it is true that record-ends won't be converted to NL-characters, so multiline text-files became a single line.

How about a command line-option?

File: apps/s_server.c [hunk 1]
```diff
@@ -966,6 +969,7 @@ const OPTIONS s_server_options[] = {
      "The number of TLSv1.3 session tickets that a server will automatically  issue" },
     {"anti_replay", OPT_ANTI_REPLAY, '-', "Switch on anti-replay protection (default)"},
     {"no_anti_replay", OPT_NO_ANTI_REPLAY, '-', "Switch off anti-replay protection"},
+    {"http_server_readmode", OPT_HTTP_SERVER_READMODE, 's', "file-opening mode for 'fopen' in -WWW and -HTTP"},
```
Comment (paulidale) [inline apps/s_server.c]: I'd prefer a _binary mode_ option rather than setting the `fopen(3)` argument on the command line.

Comment (lzsiga) [reply apps/s_server.c]: Well, I was thinking of BS2000 where open-mode can be `r,tabexp=yes` org `r,type=record,forge=seq` but maybe this is over-engineering.

(Also later one might want to turn off ebcdic-ascii conversion for binary files, in such a case a flag will be better than a string.)

File: apps/s_server.c [hunk 2]
```diff
@@ -3249,9 +3256,11 @@ static int www_body(int s, int stype, int prot, unsigned char *context)
                 break;
             }
 
-            if ((file = BIO_new_file(p, "r")) == NULL) {
+            const char *opmode = http_server_binmode==1 ? "rb" : "r";
```
Comment (levitte) [inline apps/s_server.c]: Declaration of `opmode` at the beginning of the function

**Comment by paulidale** [comment]
@mattcaswell do you need to reconfirm this?

**Comment by lzsiga** [comment]
Edited the manual.

**Comment by mattcaswell** [comment]
Since this is a new feature it can't go to 1.1.1. Is that a problem?

**Comment by paulidale** [comment]
I don't think this will cause merge problems, so master only is fine with me.

**Comment by paulidale** [comment]
@mattcaswell you need to reconfirm I think.

**Comment by paulidale** [comment]
Merged, thanks.

**Pull request closed without merging**